### PR TITLE
Issue 315: Clearable columns

### LINKED ISSUE #315: Ability to clear the contents of a column
- The UI for this feature may overload the existing `x` button on the column or it may introduce a new icon or an additional context menu.
- Deleting the contents of a column will delete the contents across all pages (if pagination is enabled)
- If a filtering expression is applied on a column of which the contents are deleted, the filter will be cleared as well.

### PR DESCRIPTION
**Pull request by Marc-Andre-Rivet**
Closes #315.

- [x] Update CHANGELOG with information about the feature and ref to this PR/issue
- [x] New tests for the feature
- [x] Visual tests for the column actions
- [ ] Documentation (`clearable` usage, action icon override through CSS & table css prop)

- Adds the ability to clear column (columns when merged) content without deleting the column itself.
- When the `clear` action is done, the associated filters are also cleared. The behavior of `delete` as been made the same as for `clear` with regard to the fitlers.
- some code rework to share filter state between FilterFactory and HeaderFactory now that header actions tie in with filtering

Re-using the currently in review changes done by @alinastarkov to have a different behavior based on whether columns are merged or not (https://github.com/plotly/dash-table/blob/f69106e5a9639ab1f7831f4cdabb0860bd1579f0/src/dash-table/utils/actions.js)

### CODE REVIEW

File: src/dash-table/derived/filter/map.ts [hunk 1]
```diff
@@ -61,20 +61,49 @@ export default memoizeOneFactory((
...
     if (value && value.length) {
-        newMap.set(safeColumnId, new SingleColumnSyntaxTree(value, column));
+        newMap.set(id, new SingleColumnSyntaxTree(value, column));
     } else {
-        newMap.delete(safeColumnId);
+        newMap.delete(id);
     }
 
     return newMap;
+}
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/derived/filter/map.ts]: Isolating the map update logic - can be called for a single column update (user edits filter) or through header `clear` and `delete` column actions (which may impact multiple columns at the same time)

File: src/dash-table/derived/filter/map.ts [hunk 2]
```diff
@@ -61,20 +61,49 @@ export default memoizeOneFactory((
...
+    const asts = Array.from(map.values());
+    const globalFilter = getMultiColumnQueryString(asts);
+
+    const rawGlobalFilter = R.map(
+        ast => ast.query || '',
+        R.filter<SingleColumnSyntaxTree>(ast => Boolean(ast), asts)
+    ).join(' && ');
+
+    setFilter(globalFilter, rawGlobalFilter, map);
+}
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/derived/filter/map.ts]: Same idea here, isolating the state update logic for both scenarios.

File: src/dash-table/derived/header/content.tsx [hunk 3]
```diff
@@ -13,15 +13,40 @@ import {
...
+        }
+    }, actions.getAffectedColumns(column, columns, columnRowIndex));
 
-function deleteColumn(column: IVisibleColumn, columns: VisibleColumns, columnRowIndex: any, setProps: SetProps, data: Data) {
-    return () => {
-        setProps(actions.deleteColumn(column, columns, columnRowIndex, data));
-    };
-}
+    clearColumnsFilter(map, affectedColumns, setFilter);
+};
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/derived/header/content.tsx]: Delete and clear apply the same logic, only the `action` changes

File: src/dash-table/derived/header/content.tsx [hunk 4]
```diff
@@ -79,14 +104,22 @@ function getSortingIcon(columnId: ColumnId, sortBy: SortBy) {
     }
 }
 
+function getColumnFlag(i: number, flag?: boolean | boolean[]): boolean {
+    return typeof flag === 'boolean' ?
+        flag :
+        !!flag && flag[i];
+}
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/derived/header/content.tsx]: renamable, clearable, deletable all use the same logic to decide whether the action is available

File: src/dash-table/derived/table/index.tsx [hunk 5]
```diff
@@ -35,12 +35,15 @@ export default (propsFn: () => ControlledTableProps) => {
...
         const props = propsFn();
 
-        return filterPropsFn(propsFn, setFilter(props.setProps, props.setState));
-    });
-    const headerFactory = new HeaderFactory(propsFn);
+        return propsAndMapFn(propsFn, setFilter(props.setProps, props.setState));
+    };
+
+    const filterFactory = new FilterFactory(augmentedPropsFn);
+    const headerFactory = new HeaderFactory(augmentedPropsFn);
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/derived/table/index.tsx]: headers now need access to map/setFilter too, generalize usage

**Comment by Marc-Andre-Rivet** [comment]
@alexcjohnson Good idea, adding a few visual tests. Keeping in mind that we have https://github.com/plotly/dash-table/issues/495 as a follow up.

**Comment by alexcjohnson** [comment]
Thanks for the visual tests!
I wonder if `ø` might look more “icon-like” than `Ø`
Also, a thought: what if we put `ø` and `×` into a css `::before` or something, so users could override these (without us having to add another prop 😏)?

**Comment by Marc-Andre-Rivet** [comment]
Action "icons" can now be overridden through css or the table's `css` prop

```
css = [
    { selector: '.column-header--clear::before', rule: `content: 'C';` },
    { selector: '.column-header--delete::before', rule: `content: 'D';` },
    { selector: '.column-header--edit::before', rule: `content: 'E';` }
]
```

**Pull request merged**